### PR TITLE
VITIS-8071 Update ReportMemory to use Hardware Contexts

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
https://jira.xilinx.com/browse/VITIS-8071
The current memory report does not correlate to how hardware contexts use memory.

If you have any opinions on the report output please let me know. I am happy to make any changes.

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
For the hardware context concept, memory is tied into a particular hardware context slot. For Alveo devices this was not an issue. However, for AIE devices with multiple hardware context slots we need a way to store and logically display these memory banks to the user.

#### How problem was solved, alternative solutions (if any) and why they were rejected
Assign to each memory bank the xclbin uuid and hw context slot from which it came. Also updated how ReportMemory displays memory banks.

#### Risks (if any) associated the changes in the commit
The output json has new fields added to each memory item. This will not interfere with any previous parsing scripts.

#### What has been tested and how, request additional testing if necessary
Ubuntu 20.04 U50C
```
dbenusov@xsjdbenusov50:/proj/rdi/staff/dbenusov$ xbutil examine -d 04:00 -r memory
WARNING: Unexpected xocl version (2.14.0) was found. Expected 2.15.0, to match XRT tools.

-------------------------------------------------
[0000:04:00.1] : xilinx_u55c_gen3x16_xdma_base_3
-------------------------------------------------

  Memory Topology
    HW Context Slot: 0
      Xclbin UUID: e106e953-cf90-4024-e075-282d1a7d820b
      Index  Tag       Type      Temp(C)  Size    Base Address
      ----------------------------------------------------------
      0      HBM[0]    MEM_HBM   37       512 MB  0x0
      1      HBM[1]    MEM_DRAM  37       512 MB  0x20000000
      2      HBM[2]    MEM_DRAM  37       512 MB  0x40000000
      3      HBM[3]    MEM_DRAM  37       512 MB  0x60000000
      4      HBM[4]    MEM_DRAM  37       512 MB  0x80000000
      5      HBM[5]    MEM_DRAM  37       512 MB  0xa0000000
      6      HBM[6]    MEM_DRAM  37       512 MB  0xc0000000
      7      HBM[7]    MEM_DRAM  37       512 MB  0xe0000000
      8      HBM[8]    MEM_DRAM  37       512 MB  0x100000000
      9      HBM[9]    MEM_DRAM  37       512 MB  0x120000000
      10     HBM[10]   MEM_DRAM  37       512 MB  0x140000000
      11     HBM[11]   MEM_DRAM  37       512 MB  0x160000000
      12     HBM[12]   MEM_DRAM  37       512 MB  0x180000000
      13     HBM[13]   MEM_DRAM  37       512 MB  0x1a0000000
      14     HBM[14]   MEM_DRAM  37       512 MB  0x1c0000000
      15     HBM[15]   MEM_DRAM  37       512 MB  0x1e0000000
      16     HBM[16]   MEM_DRAM  37       512 MB  0x200000000
      17     HBM[17]   MEM_DRAM  37       512 MB  0x220000000
      18     HBM[18]   MEM_DRAM  37       512 MB  0x240000000
      19     HBM[19]   MEM_DRAM  37       512 MB  0x260000000
      20     HBM[20]   MEM_DRAM  37       512 MB  0x280000000
      21     HBM[21]   MEM_DRAM  37       512 MB  0x2a0000000
      22     HBM[22]   MEM_DRAM  37       512 MB  0x2c0000000
      23     HBM[23]   MEM_DRAM  37       512 MB  0x2e0000000
      24     HBM[24]   MEM_DRAM  37       512 MB  0x300000000
      25     HBM[25]   MEM_DRAM  37       512 MB  0x320000000
      26     HBM[26]   MEM_DRAM  37       512 MB  0x340000000
      27     HBM[27]   MEM_DRAM  37       512 MB  0x360000000
      28     HBM[28]   MEM_DRAM  37       512 MB  0x380000000
      29     HBM[29]   MEM_DRAM  37       512 MB  0x3a0000000
      30     HBM[30]   MEM_DRAM  37       512 MB  0x3c0000000
      31     HBM[31]   MEM_DRAM  37       512 MB  0x3e0000000
      32     PLRAM[0]  MEM_DRAM  N/A      0 Byte  0x0
      33     PLRAM[1]  MEM_DRAM  N/A      0 Byte  0x0
      34     PLRAM[2]  MEM_DRAM  N/A      0 Byte  0x0
      35     PLRAM[3]  MEM_DRAM  N/A      0 Byte  0x0
      36     PLRAM[4]  MEM_DRAM  N/A      0 Byte  0x0
      37     PLRAM[5]  MEM_DRAM  N/A      0 Byte  0x0
      38     HOST[0]   MEM_DRAM  N/A      0 Byte  0x0

  DMA Transfer Metrics
    Chan[ 0].h2c:  0 Byte
    Chan[ 0].c2h:  192 Byte
    Chan[ 1].h2c:  0 Byte
    Chan[ 1].c2h:  0 Byte
```

#### Documentation impact (if any)
We may need to add  something about what a hardware context is? Not sure if we have any details in the documentation about it.

### CODE REVIEW

File: src/runtime_src/core/pcie/linux/device_linux.cpp [hunk 1]
```diff
@@ -1030,6 +1031,35 @@ struct sysfs_fcn<std::vector<VectorValueType>>
...
+      mem_data.m_type = mem_topo->m_mem_data[index].m_type;
+      mem_data.m_used = mem_topo->m_mem_data[index].m_used;
+      memcpy(mem_data.m_tag, mem_topo->m_mem_data[index].m_tag, sizeof(mem_topo->m_mem_data[index].m_tag));
+      // The following two entries are unions
+      // Within the union using any name would work. So use the first!
+      mem_data.m_size = mem_topo->m_mem_data[index].m_size;
+      mem_data.m_base_address = mem_topo->m_mem_data[index].m_base_address;
+      topology.push_back(mem_data);
+    }
+    return topology;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/device_linux.cpp:1059]: This is not really a query request, it is basically just doing formatting.   It does not belong here.

File: src/runtime_src/core/common/query_requests.h [hunk 2]
```diff
@@ -725,6 +726,32 @@ struct mem_topology_raw : request
...
+    uint8_t m_used;          // if 0 this bank is not present
+    union {
+        uint64_t m_size;     // if mem_type DDR, then size in KB;
+        uint64_t route_id;   // if streaming then "route_id"
+    };
+    union {
+        uint64_t m_base_address; // if DDR then the base address;
+        uint64_t flow_id;        // if streaming then "flow id"
+    };
+    unsigned char m_tag[16]; // DDR: BANK0,1,2,3, has to be null terminated; if streaming then stream0, 1 etc
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:744]: I don't like this.  It's a copy from xclbin.h and I don't understand why?

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/query_requests.h (hunk 2 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/query_requests.h:744]: Understood. I will move the formatting back into info_memory and remove the new query requests. We should not be doing any formatting in the query requests.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dbenusov-xilinx** [comment]
Closed this PR as it needed a rewrite to get into a usable state.

**Pull request closed without merging**